src/snap/plugin.cc

```diff
@@ -165,3 +173,243 @@ static void start_plugin(Plugin::PluginInterface* plugin, const Plugin::Meta& me
+            case Metric::Int32 : os << metric.get_int_data() << "\n";
+                break;
+            case Metric::Uint32 : os << metric.get_uint32_data() << "\n";
+                break;
+            case Metric::String : os << metric.get_string_data() << "\n";
+                break;
+            case Metric::Bool : os << metric.get_bool_data() << "\n";
+                break;
+            case Metric::Float32 : os << metric.get_float32_data() << "\n";
+                break;
+            case Metric::Float64 : os << metric.get_float64_data() << "\n";
+                break;
+            case Metric::Int64 : os << metric.get_int64_data() << "\n";
+                break;
+            case Metric::Uint64 : os << metric.get_uint64_data() << "\n";
+                break;
+        }
+    }
+    timer.print_elapsed("printCollectMetrics took ","\n");
+}
+
+Plugin::DiagnosticPrinter::Stopwatch::Stopwatch(std::ostream& os ) :
+                        os(os),
+                        started(false),
+                        stopped(false),
+                        unit("µs") {}
+
+Plugin::DiagnosticPrinter::Stopwatch::~Stopwatch() {}
+
+void Plugin::DiagnosticPrinter::Stopwatch::start() {
+    begin = std::chrono::high_resolution_clock::now();
+    started = true;
+}
+
+void Plugin::DiagnosticPrinter::Stopwatch::stop() {
+    end = std::chrono::high_resolution_clock::now();
+    stopped = true;
+}
+
+void Plugin::DiagnosticPrinter::Stopwatch::print_elapsed(std::string message_before,
+                                      std::string message_after) {
+    if (!started) {
+        std::cerr<<"Stopwatch must be started first!\n";
+        this->start();
+    }
+    if (!stopped) {
+        this->stop();
+    }
+    auto result = std::chrono::duration_cast<std::chrono::microseconds>(end - begin).count();
+    if ( result > 1000.00) {
+        result /= 1000.00;
+        unit = "ms";
+    }
+    if ( result > 1000.00) {
+        result /= 1000.00;
+        unit = "s";
+    }
+    os << message_before << result << unit << message_after;
+}
+
+std::string Plugin::DiagnosticPrinter::get_os_name() {
+    #ifdef __linux__
+    return "linux";
+    #elif _WIN32
+    return "windows 32-bit";
+    #elif _WIN64
+    return "windows 64-bit";
+    #elif __APPLE__ || __MACH__
+    return "mac OSX";
+    #else
+    return "unknown";
+    #endif
+}
+
+std::string Plugin::DiagnosticPrinter::get_architecture_name() {
+    #ifdef __i386
+    return "Intel x86";
+    #elif __amd64__
+    return "amd64";
+    #elif __ia64__
+    return "Intel Itanium";
+    #else
+    return "unknown";
+    #endif
+}
+
+void Plugin::DiagnosticPrinter::check_for_missing_requirements(const std::string& key, bool hasdefault) {
+    if (hasdefault) {
+        if (! (cfgmap.stringmap().count(key) && cfgmap.stringmap().count(key) && cfgmap.stringmap().count(key))) {
+            missing_config_requirements += "! Warning: \"" + key + "\" required by plugin and not provided in config \n";
```

Comment:
Consider adding the following:
`! Please provide config in form of: -config '{"key":"kelly", "spirit-animal":"coatimundi"}'`